### PR TITLE
feat(ext/flash): An optimized http/1.1 server

### PR DESCRIPTION
**Pull request by littledivy**
This commit adds two unstable APIs:
- "Deno.serve()"
- "Deno.serveTls()"

These APIs mirror APIs available in https://deno.land/std@0.151.0/http/server.ts?s=serve and https://deno.land/std@0.151.0/http/server.ts?s=serveTls

### CODE REVIEW

**Comment by bartlomieju** [comment]
Note to self: `console.log(request)` in a test causes a crash

**Comment by bartlomieju** [comment]
@littledivy :
- [ ] bugs with transfer encoding
- [ ] TLS support
- [ ] port remaining Hyper tests

@bartlomieju :
- [ ] unstable check on Deno.serve API
- [ ] permission check on Deno.serve API
- [ ] add Deno.serveTls API
- [ ] add “onListen” callback to make it look like std/http/server.ts
- [ ] fix typings for optional return type
- [ ] change tests to close server on failure
- [ ] fix “console.log(request)”
- [ ] debug CI timing out

File: cli/tests/unit/flash_test.ts [hunk 1]
```diff
@@ -0,0 +1,1981 @@
...
+  body: stream(""),
+  expects_chunked: true,
+  expects_con_len: false,
+});
+
+Deno.test(
+  {
+    // FIXME(bartlomieju): this test is hanging on Windows, needs to be
+    // investigated and fixed
+    ignore: Deno.build.os === "windows",
```
Comment (bartlomieju) [inline cli/tests/unit/flash_test.ts:1168]: This needs to be fixed, ignored for now to be able to land this huge PR

File: cli/dts/lib.deno.unstable.d.ts [hunk 2]
```diff
@@ -1211,12 +1211,59 @@ declare namespace Deno {
    */
   export function unrefTimer(id: number): void;
 
-  /** **UNSTABLE**: new API, yet to be vetted.
+  /**
+   * A handler for HTTP requests. Consumes a request and returns a response.
+   *
+   * Handler allows `void` or `Promise<void>` return type to enable
+   * request upgrades using `Deno.upgradeHttp()` API. It is callers responsibility
```
Comment (bartlomieju) [inline cli/dts/lib.deno.unstable.d.ts:1218]: Docs for these APIs needs to be updated, I guess we can do that in a follow up?

File: ext/ffi/lib.rs [hunk 3]
```diff
@@ -202,22 +202,23 @@ pub fn init<P: FfiPermissions + 'static>(unstable: bool) -> Extension {
       let mut work_items: Vec<PendingFfiAsyncWork> = vec![];
 
       {
-        let mut op_state = op_state_rc.borrow_mut();
-        let ffi_state = op_state.borrow_mut::<FfiState>();
+        if let Ok(mut op_state) = op_state_rc.try_borrow_mut() {
```
Comment (bartlomieju) [inline ext/ffi/lib.rs:205]: This should be reverted

File: runtime/js/90_deno_ns.js [hunk 4]
```diff
@@ -107,6 +107,8 @@
     Permissions: __bootstrap.permissions.Permissions,
     PermissionStatus: __bootstrap.permissions.PermissionStatus,
     serveHttp: __bootstrap.http.serveHttp,
+    serve: __bootstrap.flash.serve,
+    serveTls: __bootstrap.flash.serveTls,
```
Comment (bartlomieju) [inline runtime/js/90_deno_ns.js:111]: Move to unstable namespace, add check for TypeScript checks

File: runtime/ops/http.rs [hunk 5]
```diff
@@ -78,6 +82,23 @@ fn op_http_start(
   Err(bad_resource_id())
 }
 
+#[op]
+fn op_flash_upgrade_http(
```
Comment (bartlomieju) [inline runtime/ops/http.rs:86]: Why does this op has to be here?

File: runtime/worker.rs [hunk 6]
```diff
@@ -170,6 +170,7 @@ impl MainWorker {
       ops::signal::init(),
       ops::tty::init(),
       deno_http::init(),
+      deno_flash::init::<Permissions>(unstable),
```
Comment (bartlomieju) [inline runtime/worker.rs:173]: Should be exposed in workers as well

File: cli/tests/unit/flash_test.ts [hunk 7]
```diff
@@ -0,0 +1,1981 @@
...
+    }
+    {
+      const buf = new Uint8Array(1024);
+      const readResult = await conn.read(buf);
+      assert(readResult);
+      const msg = decoder.decode(buf.subarray(0, readResult));
+      assert(msg.endsWith("zar quux"));
+    }
+
+    conn.close();
```
Comment (piscisaureus) [inline cli/tests/unit/flash_test.ts:1225]: try `await conn.closeWrite()' before closing

**Review by ry — COMMENTED** [review]
Excluding the one big bundled file used in benchmarks, this patch is 4898 insertions(+), 24 deletions(-)

```
# git diff main --stat  --  ':!cli/bench/http/node_reactdom_ssr.js'
 Cargo.lock                                 |   28 +++
 Cargo.toml                                 |    3 +
 cli/bench/http/bun_http_send_file.js       |   12 +
 cli/bench/http/deno_flash_send_file.js     |   14 ++
 cli/bench/http/deno_http_flash.js          |   14 ++
 cli/bench/http/deno_http_flash_ops.js      |   33 +++
 cli/bench/http/deno_reactdom_ssr_flash.jsx |   26 ++
 cli/bench/testdata/bun_reactdom_ssr.jsx    |   23 ++
 cli/bench/testdata/deno_upgrade_http.js    |   13 +
 cli/bench/testdata/react-dom.js            |   28 +++
 cli/diagnostics.rs                         |    5 +
 cli/dts/lib.deno.unstable.d.ts             |  140 ++++++++++-
 cli/tests/unit/flash_test.ts               | 1981 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 core/lib.rs                                |    2 +
 core/resources.rs                          |    7 +
 core/runtime.rs                            |   10 +-
 ext/fetch/22_body.js                       |   16 +-
 ext/fetch/23_request.js                    |   37 ++-
 ext/fetch/23_response.js                   |    2 +-
 ext/flash/01_http.js                       |  550 ++++++++++++++++++++++++++++++++++++++++++
 ext/flash/Cargo.toml                       |   29 +++
 ext/flash/README.md                        |    7 +
 ext/flash/chunked.rs                       |  272 +++++++++++++++++++++
 ext/flash/lib.rs                           | 1451 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 ext/flash/sendfile.rs                      |   81 +++++++
 ext/http/01_http.js                        |   16 ++
 ext/http/lib.rs                            |   39 ++-
 ext/web/06_streams.js                      |    1 +
 ext/websocket/lib.rs                       |   23 +-
 runtime/Cargo.toml                         |    2 +
 runtime/build.rs                           |   10 +
 runtime/js/90_deno_ns.js                   |    2 +
 runtime/ops/http.rs                        |   23 +-
 runtime/ops/io.rs                          |   10 +
 runtime/permissions.rs                     |    9 +
 runtime/web_worker.rs                      |    1 +
 runtime/worker.rs                          |    1 +
 tools/lint.js                              |    1 +
 38 files changed, 4898 insertions(+), 24 deletions(-)
```

File: ext/flash/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1451 @@
...
+    if nread == 0 {
+      let _ = sock.write(b"0\r\n\r\n");
+      break;
+    }
+    let response = &buf[..nread];
+
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
```
Comment (ry) [inline ext/flash/lib.rs:310]: are these buffered at all? are these all separate syscalls? Presumably the return value is the number of bytes written - it's possible that the complete string isn't written?

File: ext/flash/lib.rs [hunk 9]
```diff
@@ -0,0 +1,1451 @@
...
+  }
+
+  resource.close();
+  Ok(())
+}
+
+pub struct RespondChunkedFast;
+
+impl fast_api::FastFunction for RespondChunkedFast {
+  fn function(&self) -> *const c_void {
```
Comment (ry) [inline ext/flash/lib.rs:320]: Will this be able to be removed once we have fast API supported in ops?

File: ext/flash/lib.rs [hunk 10]
```diff
@@ -0,0 +1,1451 @@
...
+      let tx = ctx.response.get(&token).unwrap();
+      tx.socket()
+    }
+  };
+
+  if let Some(response) = response {
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
```
Comment (ry) [inline ext/flash/lib.rs:378]: multiple syscalls?

Comment (piscisaureus) [reply ext/flash/lib.rs:378]: Ideally we’d use write_all_vectored() here.
Also I think we shouldn’t ignore errors.

File: ext/flash/lib.rs (hunk 9 above)
Comment (littledivy) [reply ext/flash/lib.rs:320]: yeah

File: runtime/ops/http.rs (hunk 5 above)
Comment (littledivy) [reply runtime/ops/http.rs:86]: Because op_http_upgrade is here too (circular dependency on TcpStreamResource)

**Review by ry — APPROVED** [review]
LGTM for a first pass. We'll keep iterating in main.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by rishavs** [comment]
Folks, is this PR tied to the blog post where we discussed the new Deno http server https://deno.com/blog/changes

> Deno's HTTP server is getting overhauled and we're happy to report that it is the fastest JavaScript web server ever built.

Are the benchmarks being tracked somewhere and is the data public?

**Comment by bartlomieju** [comment]
> Folks, is this PR tied to the blog post where we discussed the new Deno http server https://deno.com/blog/changes
> 
> > Deno's HTTP server is getting overhauled and we're happy to report that it is the fastest JavaScript web server ever built.
> 
> Are the benchmarks being tracked somewhere and is the data public?

https://deno.land/benchmarks#http-server-throughput

**Comment by aapoalas** [comment]
> Folks, is this PR tied to the blog post where we discussed the new Deno http server https://deno.com/blog/changes

Yes, this is the new, overhauled HTTP server. It's only for HTTP/1.1 traffic though as evident from the PR name as well.

> Are the benchmarks being tracked somewhere and is the data public?

All base benchmarks are available on the website, link to HTTP throughput specifically is here: https://deno.land/benchmarks#http-server-throughput

The latency benchmarks for the new server (flash) seem to be bugged out somehow, so those should be ignored at present.

**Comment by hayd** [comment]
> The latency benchmarks for the new server (flash) seem to be bugged out somehow, so those should be ignored at present.

Is this still the case / is it only deno_flash_send_file ?
Screenshot 2022-09-23 at 00 13 14

**Pull request merged**